libraries/lib-stretching-sequence/AudioSegment.h

```diff
@@ -0,0 +1,39 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  AudioSegment.h
+
+  Matthieu Hodgkinson
+
+**********************************************************************/
+
+#pragma once
+
+#include "SampleCount.h"
+
+#include <vector>
+
+/**
+ * @brief A generalization for audio segments, whether clips or silence between
+ * clips.
+ */
+class STRETCHING_SEQUENCE_API AudioSegment
+{
+public:
+   virtual ~AudioSegment();
+
+   /**
+    * @brief Fills `buffers` with as many as `numSamples` or the number of
+    * remaining samples, whichever is smaller.
+    * @return The number of samples actually provided.
+    */
+   virtual size_t
+   GetFloats(std::vector<float*>& buffers, size_t numSamples) = 0;
```

Comment:
You changed `float *const *` to `std::vector<float*>&` so there is information about the number of buffers.

This is all right.  But `std::vector` carries the irrelevancy about the allocator type in its second template argument, and this doesn't allow the caller simply to stack-allocate the pointers if desired.

This is why I wanted `std::span` for such uses, but that is C++20.  However it might be easily imitated in a new utility.

It's also why my preference, lacking `span`, is to pass two arguments, the array of pointers and its size.

Now should there be an explanation about which buffers are filled?  If the caller happens to pass more buffers then the width, should you expect zeroes in the extra buffers or unspecified?

This abstraction still has no interface to query the width.

None of the above is a blocking criticism.

Reply:
If we're picky, yes, we'll want `GetWidth` and specifications about `buffers.size() > GetWidth`.
`AudioSegment`s are used in `StretchingSequence`, which, like all `PlayableSequence`s in Audacity as of now, has either only mono or stereo clips. This is why a need for `GetWidth` wasn't felt in the first place.
I suggest we leave it at that for now, and address it properly when/if we ever support tracks of mixed channel counts.